src/input/type_abstract.cc

```diff
@@ -123,19 +123,22 @@ int Abstract::add_child(const Record &subrec)
 }
 
 
-bool Abstract::finish(bool is_generic) {
-	if (child_data_->finished_) return true;
+FinishStatus Abstract::finish(FinishType finish_type) {
```

Comment:
Try to merge FinishStatus and FinishType.  Introduce FinishStatus::generic. Add 
`Assert( finish_type!=FinishStatus::none)`.